dagql/objects.go

```diff
@@ -655,18 +656,32 @@ func (r ObjectResult[T]) call(
 ) (AnyResult, error) {
 	ctx = idToContext(ctx, newID)
 	ctx = srvToContext(ctx, s)
-	callCacheKey := newID.Digest()
-	if doNotCache {
-		callCacheKey = ""
-	}
-
 	var opts []CacheCallOpt
 	if s.telemetry != nil {
 		opts = append(opts, WithTelemetry(func(ctx context.Context) (context.Context, func(AnyResult, bool, error)) {
 			return s.telemetry(ctx, r, newID)
 		}))
 	}
-	res, err := s.Cache.GetOrInitializeWithCallbacks(ctx, callCacheKey, true, func(ctx context.Context) (*CacheValWithCallbacks, error) {
+
+	cacheKey := cache.CacheKey[CacheKeyType]{
+		ResultKey: string(newID.Digest()),
+	}
+	if doNotCache {
+		cacheKey.ResultKey = ""
+	}
+	// dedupe concurrent calls only if the ID digest is the same and if the two calls are from the same session
+	// we don't want to dedupe across sessions since:
+	// 1. it creates problems when one session closes and others were waiting on the result
+	// 2. it makes it easy to accidentally leak session specific information that isn't yet precisely scoped in the ID
```

Comment:
Is client id instead of session id preferable?

For example, I could imagine a case where a nested client (like a function call) could close early, etc.

Reply:
True, will swap that out